internal/server/v2/observation/calculation_util.go

```diff
@@ -212,90 +211,143 @@ func compareFacet(facet1, facet2 *pb.Facet) bool {
 		}
 	}
+	return result
 }
 
-// Combine two sets of candidate observations using an operator token.
-func evalBinaryExpr(
-	x, y map[string]map[string][]*pb.PointStat,
+// Combine two PointStat series using an operator token.
+func mergePointStat(
+	x, y []*pb.PointStat,
 	op token.Token,
-) (map[string]map[string][]*pb.PointStat, error) {
-	result := map[string]map[string][]*pb.PointStat{}
-	for entity, xFacetObs := range x {
-		newFacetObs := map[string][]*pb.PointStat{}
-		for facetId, xObs := range xFacetObs {
-			yFacetObs, ok := y[entity]
-			if !ok {
-				continue
-			}
-			yObs, ok := yFacetObs[facetId]
-			if !ok {
-				continue
+) ([]*pb.PointStat, error) {
+	result := []*pb.PointStat{}
+	xIdx, yIdx := 0, 0
+	for xIdx < len(x) && yIdx < len(y) {
+		xDate, yDate := x[xIdx].GetDate(), y[yIdx].GetDate()
+		if xDate < yDate {
+			xIdx++
+		} else if yDate < xDate {
+			yIdx++
+		} else {
+			xVal := x[xIdx].GetValue()
+			yVal := y[yIdx].GetValue()
+			var val float64
+			switch op {
+			case token.ADD:
+				val = xVal + yVal
+			case token.SUB:
+				val = xVal - yVal
+			case token.MUL:
+				val = xVal * yVal
+			case token.QUO:
+				if yVal == 0 {
+					return nil, fmt.Errorf("denominator cannot be zero")
+				}
+				val = xVal / yVal
+			default:
+				return nil, fmt.Errorf("unsupported op (token) %v", op)
 			}
-			newObs := []*pb.PointStat{}
-			xIdx, yIdx := 0, 0
-			for xIdx < len(xObs) && yIdx < len(yObs) {
-				xDate, yDate := xObs[xIdx].GetDate(), yObs[yIdx].GetDate()
-				if xDate < yDate {
-					xIdx++
-				} else if yDate < xDate {
-					yIdx++
-				} else {
-					xVal := xObs[xIdx].GetValue()
-					yVal := yObs[yIdx].GetValue()
-					var val float64
-					switch op {
-					case token.ADD:
-						val = xVal + yVal
-					case token.SUB:
-						val = xVal - yVal
-					case token.MUL:
-						val = xVal * yVal
-					case token.QUO:
-						if yVal == 0 {
-							return nil, fmt.Errorf("denominator cannot be zero")
+			result = append(result, &pb.PointStat{
+				Date:  xDate,
+				Value: proto.Float64(val),
+			})
+			xIdx++
+			yIdx++
+		}
+	}
+	return result, nil
+}
+
+// Combine two ObservationResponses using an operator token.
+func evalBinaryExpr(
+	x, y *pbv2.ObservationResponse,
+	op token.Token,
+) (*pbv2.ObservationResponse, error) {
+	result := &pbv2.ObservationResponse{
+		// Intermediate responses have no StatVar, so use a placeholder.
+		ByVariable: map[string]*pbv2.VariableObservation{"": {}},
+		Facets:     map[string]*pb.Facet{},
+	}
+	if len(x.ByVariable) != 1 || len(y.ByVariable) != 1 {
+		return nil, fmt.Errorf("more than one variable in intermediate response")
+	}
+	for _, xVariableObs := range x.ByVariable {
+		for _, yVariableObs := range y.ByVariable {
+			for entity, xEntityObs := range xVariableObs.ByEntity {
+				yEntityObs, ok := yVariableObs.ByEntity[entity]
+				if !ok {
+					continue
+				}
+				newOrderedFacets := []*pbv2.FacetObservation{}
+				for i := 0; i < len(xEntityObs.OrderedFacets); i++ {
```

Comment:
after line 314, can define two short-name alias and use them below

xFacets := xEntityObs.OrderedFacets
yFacets := yEntityObs.OrderedFacets

Reply:
Done